File: prog1/Lex.java
//Lex.java
//Aaron Hom
//PA1 
//Takes input and output files
//Arranges contents of input file in alphabetical order
//Prints to output file

import java.io.*;
import java.util.Scanner;
import java.io.FileWriter;
import java.io.PrintWriter;

class Lex{

    public static void main(String[] args) throws IOException{

	Scanner in = null;
	PrintWriter out = null;
	String lines = null;
	String[] inFile = null;
	int lineNum = 0;

	if(args.length != 2){ //checks if there are two arguments 
	    System.err.println("Wrong number of arguments");
	    System.exit(1);
	}

	in = new Scanner(new File(args[0])); //opens scanner

	int countLines = 0; //variable to count lines
	while(in.hasNextLine()){ //loop to count lines
	    countLines++; //add 1 to line total
	    in.nextLine(); //move to next line
	}
	in.close(); //close scanner
	in = null;

	in = new Scanner(new File(args[0])); //open another scanner
	out = new PrintWriter(new FileWriter(args[1])); //open output printer
	List A = new List(); //initializes arrays
	inFile = new String[countLines]; //string array to keep track of lines

	while(in.hasNextLine()){ //copies lines of file into string array
	    inFile[lineNum++] = in.nextLine(); //gives each line an index
	}

	//This is where it kind of gets like insertion sort
        A.append(0); //place first line from array in as first element
	for(int i = 1; i < inFile.length; i++){ 
	    String words = inFile[i]; 
	    int j = i-2;
	    A.moveBack();
	    while(j >= 0 && words.compareTo(inFile[A.get()]) <= 0){ //compares lines
		j--; 
		A.movePrev(); //moves farther back if not in alphabetical place
	    }
	    if(A.index() >= 0){
		A.insertAfter(i); //inserts line after it finds alphabetical place
 	    }else{
		A.prepend(i); //line gets inserted in back if outside of index
	    }
	}
	A.moveFront(); //moves cursor back to front
	while(A.index() >= 0){ //loop to print array in order
	    out.println(inFile[A.get()]);
	    A.moveNext();
	}
	in.close(); //end
	out.close();
	}
    }

File: prog2/Lex.c
//Lex.c
//Aaron Hom
//pa2
//Takes in an input file, sorts lines in alphabetical order, prints to out file

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"
#define MAX_LEN 500

int main(int argc, char* argv[]){
  int numLines = -1;
  FILE *in, *out;
  char words[MAX_LEN];

  if(argc != 3){ //error message if wrong number of arguments
    printf("Wrong number of arguments");
    exit(1);
  }

  in = fopen(argv[1], "r"); //files for reading and writing
  out = fopen(argv[2], "w");

  if(in == NULL){
    printf("File for reading not found"); //error checking
    exit(1);
  }
  if(out == NULL){
    printf("File for writing not found");
    exit(1);
  }

  while(fgets(words, MAX_LEN, in) != NULL){ //counts strings 
    numLines++;
  }

  rewind(in); //set back to start

  char line[numLines][MAX_LEN];
  int tempNumLines = 0;

  while(fgets(words, MAX_LEN, in) != NULL){ //copies lines of in file
    strcpy(line[tempNumLines++], words);
  }

  List A = newList(); //creates new list
  append(A, 0); //inserts first element in list
 
  for(int i = 1; i < numLines+1 ; i++){ //places strings in their appropriate place
    char *tempLine = line[i];
    int j = i-2;
    moveBack(A);
    while(j >= 0 && strcmp(tempLine, line[get(A)]) <= 0){
      j--;
      movePrev(A);
    }
    if(index(A) >= 0){
      insertAfter(A, i);
    }else{
      prepend(A, i);
    }
  }
  moveFront(A);

  while(index(A) >= 0){ //prints out ordered strings from out file
    fprintf(out, "%s\n", line[get(A)]);
    moveNext(A);
  }

  fclose(in); //closes files for reading and writing
  fclose(out);

  freeList(&A); //frees memory
  return(0);
}
